packages/openapi-to-graphql/test/file_upload.test.ts

```diff
@@ -0,0 +1,173 @@
+      if (Oas3Tools.isHttpMethod(method) && method !== 'get') oasMutCount++
+    }
+  }
+  const gqlTypes = Object.keys(createdSchema._typeMap.Mutation.getFields())
+        .length
+  expect(gqlTypes).toEqual(oasMutCount)
+})
+
+test('registers the graphql-upload Upload scalar type', async () => {
+  const query = `{
+    __type(name: "Upload") {
+      name
+      kind
+    }
+  }`
+
+  const result = await graphql(createdSchema, query)
+  expect(result).toEqual({
+    data: {
+      __type: {
+        name: 'Upload',
+        kind: 'SCALAR'
+      }
+    }
+  })
+})
+
+test('introspection for mutations returns a mutation matching the custom field specified for the multipart API definition', async () => {
+  const query = `{
+    __schema {
+      mutationType {
+        fields {
+          name
+          args {
+            name
+            type {
+              name
+              kind
+            }
+          }
+          type {
+            name
+            kind
+          }
+        }
+      }
+    }
+  }`
+
+  const result = await graphql(createdSchema, query)
+
+  expect(result).toEqual({
+    data: {
+      __schema: {
+        mutationType: {
+          fields: expect.arrayContaining([
+            expect.objectContaining({
+              name: 'fileUploadTest',
+              args: expect.arrayContaining([
+                expect.objectContaining({
+                  name: 'uploadInput'
+                })
+              ])
+            })
+          ])
+        }
+      }
+    }
+  })
+})
+
+test('upload completes without any error', async () => {
+  // setup graphql for integration test
+  const graphqlServer = createServer(async (req, res) => {
+    try {
+      const operation = await processRequest(req, res) as GraphQLOperation
+      const result = await graphql(createdSchema, operation.query, null, null, operation.variables)
+      res.end(JSON.stringify(result))
+    } catch (e) {
+      console.log(e)
+    }
+  })
+
+  const { port: graphqlServerPort, close: closeGraphQLServer } = await new Promise((resolve, reject) => {
+    graphqlServer.listen(function (err) {
+      if (err) {
+        return reject(err)
+      }
+
+      return resolve({
+        port: this.address().port,
+        close: () => this.close()
+      })
+    })
+  })
+
+  const vol = new Volume()
+
+  // create mocked in memory file for upload
+  vol.fromJSON({
+    './README.md': '1'
+  }, '/app')
+
+  // prepare request to match graphql multipart request spec
+  // https://github.com/jaydenseric/graphql-multipart-request-spec
+  const form = new FormData()
+  const query = `
+    mutation FileUploadTest($file: Upload!) {
+      fileUploadTest(uploadInput: { file: $file }) {
+        id
+        url
+      }
+    }
+  `
+  form.append('operations', JSON.stringify({ query, variables: { file: null } }))
+  form.append('map', JSON.stringify({ 0: ['variables.file'] }))
+  form.append('0', vol.createReadStream('/app/README.md'), {
```

Comment:
I am not very experienced using `graphql-upload` and creating GraphQL queries this way. Can you clarify how this all works? The `file: null`, the `'map'`, and the `'0'`? Otherwise, I'm very happy with everything else.

Reply:
Hey @Alan-Cha this is the internals for how the implementation of the specification I adopted to support file upload, it's outline in details here https://github.com/jaydenseric/graphql-multipart-request-spec . A typical user wouldn't need to do this manually there's a number of  libraries (including one for Apollo) that already implement this for the client.

Reply:
Okay, I see now. Thanks for clarifying that!